### PR TITLE
Open interaction dialog when adding prep in ExchangeOut

### LINKED ISSUE #922: Exchange Out preparations does not trigger Add Items dialog
Added by @grantfitzsimmons on Nov 1, 2023:

> ## Specify 7 Behavior
> 
> You can create an **Exchange Out**, but the **Total Preps** and **Total Items** do not work (https://github.com/specify/specify7/issues/1193)
>  
> image
> 
> Most importantly, **Exchange Out Preps** _does not work!_
> This means you can't create new preparations or link them with existing records.
> 
> ## Specify 6 Behavior
> 
> 1. Go to **Interactions**
> 2. Click on **Exchange Out**
>    image
> 4. See this dialog:
> 	image
> 	a. Enter Items
>    image
> 	b. Record Set
> 	image
> 6. Create the **Exchange Out**
>    image
> 7. See new record
> image
> 
> 
> 

---

Currently, clicking on the respective option in the Interactions dialog, triggers the following alert messages:

Screen Shot 2021-08-29 at 5 03 27 PM

Screen Shot 2021-08-29 at 5 03 36 PM

Screen Shot 2021-08-29 at 4 53 10 PM

See also https://github.com/specify/specify7/issues/516

### PR DESCRIPTION
**Pull request by CarolineDenis**
Fixes #922

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [x] Add relevant issue to release milestone

### Testing instructions

- Go to interactions 
- Open Exchange Out 
- Add a prep
- [ ] verify the interaction dialog opens
- [ ] verify that you can add a prep by selecting "By choosing a record set" 
- [ ] verify that you can add a prep by selecting "By entering catalog number" 
- [ ] verify that you can add an unassociated item 
- [ ] test adding a preparation for ExchangeIn

### CODE REVIEW

File: specifyweb/frontend/js_src/lib/components/FormSliders/IntegratedRecordSelector.tsx [hunk 1]
```diff
@@ -148,7 +152,10 @@ export function IntegratedRecordSelector({
                 actionTable={collection.related.specifyTable}
                 itemCollection={
                   collection as Collection<
-                    DisposalPreparation | GiftPreparation | LoanPreparation
+                    | DisposalPreparation
+                    | ExchangeOutPrep
+                    | GiftPreparation
+                    | LoanPreparation
                   >
```
Comment (melton-jason) [inline specifyweb/frontend/js_src/lib/components/FormSliders/IntegratedRecordSelector.tsx:159]: replace with 
```ts
collection as Collection<AnyInteractionPreparation>
```

File: specifyweb/frontend/js_src/lib/components/Interactions/InteractionDialog.tsx [hunk 2]
```diff
@@ -58,7 +59,7 @@ export function InteractionDialog({
   readonly actionTable: SpecifyTable;
   readonly isLoanReturn?: boolean;
   readonly itemCollection?: Collection<
-    DisposalPreparation | GiftPreparation | LoanPreparation
+    DisposalPreparation | ExchangeOutPrep | GiftPreparation | LoanPreparation
```
Comment (melton-jason) [inline specifyweb/frontend/js_src/lib/components/Interactions/InteractionDialog.tsx:62]:
```suggestion
    AnyInteractionPreparation
```

File: specifyweb/frontend/js_src/lib/components/Interactions/PrepDialog.tsx [hunk 3]
```diff
@@ -38,9 +40,9 @@ export function PrepDialog({
 }: {
   readonly onClose: () => void;
   readonly preparations: RA<PreparationData>;
-  readonly table: SpecifyTable<Disposal | Gift | Loan>;
+  readonly table: SpecifyTable<Disposal | ExchangeOut | Gift | Loan>;
   readonly itemCollection?: Collection<
-    DisposalPreparation | GiftPreparation | LoanPreparation
+    DisposalPreparation | ExchangeOutPrep | GiftPreparation | LoanPreparation
```
Comment (melton-jason) [inline specifyweb/frontend/js_src/lib/components/Interactions/PrepDialog.tsx:45]:
```suggestion
    AnyInteractionPreparation
```

File: specifyweb/frontend/js_src/lib/components/Interactions/PrepDialog.tsx [hunk 4]
```diff
@@ -152,10 +154,15 @@ export function PrepDialog({
...
+          const tableName =
+            table.name === 'ExchangeOut'
+              ? `${table.name}Prep`
+              : `${table.name}Preparation`;
+          const itemTable = strictGetTable(tableName) as SpecifyTable<
+            | DisposalPreparation
+            | ExchangeOutPrep
+            | GiftPreparation
+            | LoanPreparation
           >;
```
Comment (melton-jason) [inline specifyweb/frontend/js_src/lib/components/Interactions/PrepDialog.tsx:166]: Instead of hardcoding the table name here, it can be fetched dynamically using the constant defined in an above comment (a tuple containing the name of all InteractionPreparations): 

```suggestion
          const itemTable = defined(
            table.relationships.find((relationship) =>
              interactionPrepTables.includes(relationship.relatedTable.name)
            )?.table
          ) as SpecifyTable<AnyInteractionPreparation>;
```

File: specifyweb/frontend/js_src/lib/components/Interactions/PrepDialog.tsx [hunk 5]
```diff
@@ -192,6 +199,10 @@ export function PrepDialog({
               'disposalPreparations',
               items as RA<SpecifyResource<DisposalPreparation>>
             );
+            toTable(interaction, 'ExchangeOut')?.set(
+              'exchangeOutPreps',
+              items as RA<SpecifyResource<ExchangeOutPrep>>
+            );
```
Comment (melton-jason) [inline specifyweb/frontend/js_src/lib/components/Interactions/PrepDialog.tsx:205]: The following lines can be approached dynamically, rather than checking for each individual interaction: 
https://github.com/specify/specify7/blob/a2136ba121d518d4ae7db35b22b2dd6fc26aaa87/specifyweb/frontend/js_src/lib/components/Interactions/PrepDialog.tsx#L187-L205 

But first, we need to find a way to specify a type of all the base Interaction tables which have associated TableNamePreparation relationships. 

We would preferably find some way to extract the value directly from the `Tables` type, but I was unable to find a clean solution. 

Instead, I use the type `AnyInteractionPreparation` to get the possible table names of the InteractionPreparations table and remove the `Prep*` suffix, extracting the prefix without the preparation part. 

```ts
type ExtractInteraction<T extends string> =
  T extends `${infer Prefix}Prep${string}` ? Prefix : never;

export type InteractionsWithPreps = Tables[ExtractInteraction<
  AnyInteractionPreparation['tableName']
>];
```

Now, any place which needs the Interaction part of an AnyInteractionPreparation, we can replace with the `InteractionsWithPreps` preps. 

---

Instead of the `toTable` check for each interaction and setting the hardcoded field name to the value of the items, a helper function can be defined to handle the general case: 

```ts
function setPreparationItems(
  interaction: SpecifyResource<InteractionsWithPreps>,
  items: RA<SpecifyResource<AnyInteractionPreparation>>
): void {
  const preparationField = interaction.specifyTable.relationships.find(
    (relationship) =>
      interactionPrepTables.includes(relationship.relatedTable.name)
  );
  interaction.set(preparationField?.name, items);
}
```

Now all of the `toTable(...).set(somePrep, items)` can be removed and replaced with a single `setPreparationItems(interaction, items)`

**Review by melton-jason — CHANGES_REQUESTED** [review]
Currently, `ExchangeOut` is not the only Interaction missing a dialog: `ExchangeIn` is missing one as well! 

I have some suggestions to make our current approach more generic without relying on hardcoding the tables which are supported.

File: specifyweb/frontend/js_src/lib/components/Interactions/InteractionsDialog.tsx [hunk 6]
```diff
@@ -34,8 +34,8 @@ export const interactionWorkflowTables: Partial<
   Loan: 'CollectionObject',
   Gift: 'CollectionObject',
   Disposal: 'CollectionObject',
+  ExchangeOut: 'Preparation',
```
Comment (melton-jason) [inline specifyweb/frontend/js_src/lib/components/Interactions/InteractionsDialog.tsx:37]: Were the values of this object going to be used anywhere? 
Currently, only the key values are used. 

With little effort this can functionally be replaced with a new constant for values in `InteractionsWithPreps` defined the same way as `interactionPrepTables`.

**Review by melton-jason — APPROVED** [review]

**Review by emenslin — CHANGES_REQUESTED** [review]
### Testing instructions

- Go to interactions 
- Open Exchange Out 
- Add a prep
- [x] verify the interaction dialog opens
- [ ] verify that you can add a prep by selecting "By choosing a record set" 
- [ ] verify that you can add a prep by selecting "By entering catalog number" 
- [ ] verify that you can add an unassociated item 
- [ ] test adding a preparation for ExchangeIn

There is no option for adding an unassociated item when creating an exchange in/out, only in the add items dialog. 
Screenshot 2024-05-13 102428
Screenshot 2024-05-13 102444



Also although you can add preparations through record set or catalog numbers in the create exchange in/out, when you try to add any items after that it just adds a blank preparation

https://github.com/specify/specify7/assets/150940840/ec987a94-c399-41bf-b166-118c01281a08


Basically, the initial dialog works aside from missing the add unassociated preps button but the add items dialog will only add blank preparations. 

Also side note, the exchange in default form does not include the preparations subview, is there a reason for that?

**Comment by melton-jason** [comment]
> ### Testing instructions
> * Go to interactions
> * Open Exchange Out
> * Add a prep
> * [x]  verify the interaction dialog opens
> * [ ]  verify that you can add a prep by selecting "By choosing a record set"
> * [ ]  verify that you can add a prep by selecting "By entering catalog number"
> * [ ]  verify that you can add an unassociated item
> * [ ]  test adding a preparation for ExchangeIn
> 
> There is no option for adding an unassociated item when creating an exchange in/out, only in the add items dialog. Screenshot 2024-05-13 102428 Screenshot 2024-05-13 102444
> 
> Also although you can add preparations through record set or catalog numbers in the create exchange in/out, when you try to add any items after that it just adds a blank preparation
> 
>  LK8UmhhIet.mp4 
> Basically, the initial dialog works aside from missing the add unassociated preps button but the add items dialog will only add blank preparations.
> 
> Also side note, the exchange in default form does not include the preparations subview, is there a reason for that?

@emenslin 

Both of these issues have been fixed. 
- All Interactions with Preparations (Disposal, ExchangeIn, ExchangeOut, Gift, Loan) should now have the `Without Preparations` button when creating the Interaction. 
  - Previously only Gift and Loan had the button
- When adding Preparations from the Interaction, it should no longer add empty Interactions.

**Review by emenslin — APPROVED** [review]
### Testing instructions

- Go to interactions 
- Open Exchange Out 
- Add a prep
- [x] verify the interaction dialog opens
- [x] verify that you can add a prep by selecting "By choosing a record set" 
- [x] verify that you can add a prep by selecting "By entering catalog number" 
- [x] verify that you can add an unassociated item 
- [x] test adding a preparation for ExchangeIn

Looks good, the without preparation button is there and it adds items correctly now!

**Review by alesan99 — APPROVED** [review]
- [x] verify the interaction dialog opens
- [x] verify that you can add a prep by selecting "By choosing a record set" 
- [x] verify that you can add a prep by selecting "By entering catalog number" 
- [ ] verify that you can add an unassociated item 
- [ ] test adding a preparation for ExchangeIn

Checks out 👍
I also found it weird that there is no preparation subview in the default Exchange In form, but perhaps that can be addressed elsewhere
